File: lib/src/prototime.dart
import 'package:fixnum/fixnum.dart' as $fixnum;
import 'package:qtya_proto/src/generated/qtya/core/v1/prototime.pb.dart';

/// Extension on proto date util to convert back and forth the time;
extension ProtoTimeConverter on ProtoTime {
  /// Converts this object to DateTime
  DateTime toTime() {
    return DateTime.fromMillisecondsSinceEpoch(unixmillis.toInt());
  }
}

/// Extension to convert DateTime to proto
extension DateTimeProtoExtension on DateTime {
  /// FromTime converts proto time from DateTime instance;
  ProtoTime toProto() {
    return ProtoTime(unixmillis: $fixnum.Int64(millisecondsSinceEpoch));
  }
}

File: qtya/core/v1/prototime.go
package corev1

import "time"

func FromTimePtr(t *time.Time) *ProtoTime {
	if t == nil {
		return nil
	}

	return &ProtoTime{
		Unixmillis: uint64(t.UnixMilli()),
	}
}

func FromTime(t time.Time) *ProtoTime {
	return &ProtoTime{
		Unixmillis: uint64(t.UnixMilli()),
	}
}

func (pt *ProtoTime) ToTime() time.Time {
	return time.UnixMilli(int64(pt.GetUnixmillis()))
}
